moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp

```diff
@@ -350,14 +393,23 @@ bool JogCalcs::cartesianJogCalcs(geometry_msgs::TwistStamped& cmd, trajectory_ms
     Eigen::Vector3d translation_vector(cmd.twist.linear.x, cmd.twist.linear.y, cmd.twist.linear.z);
     Eigen::Vector3d angular_vector(cmd.twist.angular.x, cmd.twist.angular.y, cmd.twist.angular.z);
 
-    // We solve (planning_frame -> base -> cmd.header.frame_id)
-    // by computing (base->planning_frame)^-1 * (base->cmd.header.frame_id)
-    const auto tf_planning_to_cmd_frame =
-        kinematic_state_->getGlobalLinkTransform(parameters_.planning_frame).inverse() *
-        kinematic_state_->getGlobalLinkTransform(cmd.header.frame_id);
-
-    translation_vector = tf_planning_to_cmd_frame.linear() * translation_vector;
-    angular_vector = tf_planning_to_cmd_frame.linear() * angular_vector;
+    // If the incoming frame is empty or is the command frame, we use the previously calculated tf
+    if (cmd.header.frame_id.empty() || cmd.header.frame_id == parameters_.robot_link_command_frame)
+    {
+      translation_vector = tf_moveit_to_robot_cmd_frame_.linear() * translation_vector;
+      angular_vector = tf_moveit_to_robot_cmd_frame_.linear() * angular_vector;
+    }
+    else
+    {
+      // We solve (planning_frame -> base -> cmd.header.frame_id)
+      // by computing (base->planning_frame)^-1 * (base->cmd.header.frame_id)
+      const auto tf_moveit_to_incoming_cmd_frame =
+          kinematic_state_->getGlobalLinkTransform(parameters_.planning_frame).inverse() *
+          kinematic_state_->getGlobalLinkTransform(cmd.header.frame_id);
```

Comment:
These transforms were previously calculated around L231. It looks like `tf_moveit_to_robot_cmd_frame_` is not used around line 231, though. Can you move L231 here? Then delete this local variable `tf_moveit_to_incoming_cmd_frame`

Reply:
These are 2 different transformations:

- `tf_moveit_to_robot_cmd_frame_` which is the moveit planning frame -> robot command frame (specified in yaml file) transformation. I believe it was calculated early (here in L231) so that it would be updated and available if the rest of the jogging process failed / ended early, and would always be available for the C++ API (via `getCommandFrameTransform()`) regardless of jogging status
- `tf_moveit_to_incoming_cmd_frame` is the moveit planning frame -> incoming twist command frame transformation and is calculated and used in the Cartesian jogging

We only use one of these to actually jog every cycle. If the twist command frame is empty, or it's in the same frame given as the "robot command frame" in the parameters, we use the already calculated `tf_moveit_to_robot_cmd_frame_`. If it is a different frame, we jog in that frame and calculate the (local) `tf_moveit_to_incoming_cmd_frame`

I believe this was the desired intent before this PR, and the changes here are to avoid recalculating `tf_moveit_to_robot_cmd_frame_` if the jogging command is in that frame. I can discuss more, but I think this is OK

Reply:
ok, that sounds good. A comment at Line 230 would be good:

// Calculate this transform to ensure it's available via C++ API

Reply:
Sounds good, see new commit